src/main/java/com/google/changestreams/sample/bigquery/changelog/fullschema/SpannerChangeStreamsToBigQuery.java

```diff
@@ -0,0 +1,329 @@
+import com.google.cloud.teleport.v2.coders.FailsafeElementCoder;
+import com.google.cloud.teleport.v2.transforms.DLQWriteTransform;
+import com.google.cloud.teleport.v2.values.FailsafeElement;
+import org.apache.beam.runners.dataflow.options.DataflowPipelineOptions;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Write;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Write.CreateDisposition;
+import org.apache.beam.sdk.io.gcp.bigquery.InsertRetryPolicy;
+import org.apache.beam.sdk.io.gcp.bigquery.WriteResult;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerConfig;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerIO;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.model.DataChangeRecord;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.Flatten;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Set;
+
+public class SpannerChangeStreamsToBigQuery {
+
+  /**
+   * String/String Coder for FailsafeElement.
+   */
+  public static final FailsafeElementCoder<String, String> FAILSAFE_ELEMENT_CODER =
+    FailsafeElementCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of());
+
+  private static final Logger LOG = LoggerFactory.getLogger(SpannerChangeStreamsToBigQuery.class);
+
+  /**
+   * Main entry point for executing the pipeline.
+   *
+   * @param args The command-line arguments to the pipeline.
+   */
+  public static void main(String[] args) {
+    LOG.info("Starting to replicate change records from Spanner Change Streams to BigQuery");
+
+    Options options = PipelineOptionsFactory.fromArgs(args).withValidation().as(Options.class);
+
+    options.setStreaming(true);
+    options.setEnableStreamingEngine(true);
+
+    validateOptions(options);
+    run(options).waitUntilFinish();
+  }
+
+  private static void validateOptions(Options options) {
+    final String startTimestampStr = options.getStartTimestamp();
+    final Timestamp startTimestamp = Timestamp.parseTimestamp(startTimestampStr);
+    final String endTimestampStr = options.getEndTimestamp();
+    if (!endTimestampStr.isEmpty()) {
+      final Timestamp endTimestamp = Timestamp.parseTimestamp(endTimestampStr);
+      if (startTimestamp.compareTo(endTimestamp) <= 0) {
+        throw new IllegalArgumentException(
+          "startTimestamp must be smaller than or equal to endTimestamp.");
```

Comment:
If start timestamp can be equal to the end timestamp we should change the comparison to `startTimestamp.compareTo(endTimestamp) < 0`.

Reply:
Good catch, done.